drv/stm32h7-update-server/src/main.rs

```diff
@@ -0,0 +1,342 @@
+
+        ringbuf_entry!(Trace::FinishStart);
+        if flash.optsr_cur().read().swap_bank_opt().bit() {
+            flash
+                .optsr_prg()
+                .modify(|_, w| w.swap_bank_opt().clear_bit());
+        } else {
+            flash.optsr_prg().modify(|_, w| w.swap_bank_opt().set_bit());
+        }
+
+        flash.optcr().modify(|_, w| w.optstart().set_bit());
+
+        loop {
+            if !flash.optsr_cur().read().opt_busy().bit() {
+                break;
+            }
+        }
+
+        ringbuf_entry!(Trace::FinishEnd);
+        Ok(())
+    }
+
+    fn poll_flash_done(&mut self) -> Result<(), RequestError<UpdateError>> {
+        let mut cnt = 0;
+        let mut seen = false;
+
+        loop {
+            if self.flash.bank2().sr.read().qw().bit() {
+                seen = true;
+                cnt = 0;
+            }
+
+            if seen && !self.flash.bank2().sr.read().qw().bit() {
+                break;
+            } else {
+                cnt += 1;
+            }
+
+            if cnt == FLASH_TIMEOUT {
+                break;
+            }
+        }
+
+        if cnt == FLASH_TIMEOUT {
+            return Err(UpdateError::Timeout.into());
+        }
+
+        if self.flash.bank2().sr.read().dbeccerr().bit() {
+            return Err(UpdateError::EccDoubleErr.into());
+        }
+
+        if self.flash.bank2().sr.read().sneccerr1().bit() {
+            return Err(UpdateError::EccSingleErr.into());
+        }
+
+        if self.flash.bank2().sr.read().rdserr().bit() {
+            return Err(UpdateError::SecureErr.into());
+        }
+
+        if self.flash.bank2().sr.read().rdperr().bit() {
+            return Err(UpdateError::ReadProtErr.into());
+        }
+
+        if self.flash.bank2().sr.read().operr().bit() {
+            return Err(UpdateError::WriteEraseErr.into());
+        }
+
+        if self.flash.bank2().sr.read().incerr().bit() {
+            return Err(UpdateError::InconsistencyErr.into());
+        }
+
+        if self.flash.bank2().sr.read().strberr().bit() {
+            return Err(UpdateError::StrobeErr.into());
+        }
+
+        if self.flash.bank2().sr.read().pgserr().bit() {
+            return Err(UpdateError::ProgSeqErr.into());
+        }
+
+        if self.flash.bank2().sr.read().wrperr().bit() {
+            return Err(UpdateError::WriteProtErr.into());
+        }
+
+        Ok(())
+    }
+
+    // RM0433 Rev 7 section 4.3.9
+    // Following Single write sequence
+    fn write_word(
+        &mut self,
+        word_number: usize,
+        bytes: &[u8],
+    ) -> Result<(), RequestError<UpdateError>> {
+        ringbuf_entry!(Trace::WriteStart);
+        self.flash.bank2().cr.write(|w| w.pg().set_bit());
+
+        if word_number > BANK_WORD_LIMIT {
+            return Err(UpdateError::OutOfBounds.into());
+        }
+
+        let start = BANK_ADDR + (word_number * FLASH_WORD_BYTES) as u32;
+
+        if bytes.len() != FLASH_WORD_BYTES {
```

Comment:
(If `bytes` had type `&[u8; FLASH_WORD_BYTES]` you could move this check to compile time.)